http/src/main/java/io/micronaut/http/filter/FilterRunner.java

```diff
@@ -1,64 +1,316 @@
+            } else if (filter instanceof InternalFilter.TerminalReactive || filter instanceof InternalFilter.Terminal) {
+                ExecutionFlow<? extends HttpResponse<?>> terminalFlow;
+                if (filter instanceof InternalFilter.Terminal t) {
+                    try {
+                        terminalFlow = t.responseFlow().apply(request);
+                    } catch (Throwable e) {
+                        terminalFlow = ExecutionFlow.error(e);
+                    }
+                } else {
+                    terminalFlow = ReactiveExecutionFlow.fromPublisher(((InternalFilter.TerminalReactive) filter).responsePublisher());
+                }
+                // this is almost never available immediately, so don't bother with asDone checks
+                terminalFlow.onComplete((resp, fail) -> {
+                    response = resp;
+                    failure = fail;
+                    responseNeedsProcessing = true;
+                    index--;
+                    workResponse();
+                });
+                // request work is done
+                return;
+            } else {
+                throw new IllegalStateException("Unknown filter type");
+            }
+        }
+    }
+
+    private void workResponse() {
+        while (true) {
+            if (responseNeedsProcessing) {
+                responseNeedsProcessing = false;
+                if (!processResponse0()) {
+                    // suspend
+                    return;
+                }
+            }
+
+            index--;
+
+            if (responseSuspensionPoint.filterIndex == index) {
+                SuspensionPoint<HttpResponse<?>> suspensionPoint = responseSuspensionPoint;
+                responseSuspensionPoint = suspensionPoint.next;
+                boolean completed;
+                if (failure == null) {
+                    completed = suspensionPoint.complete(response);
+                } else {
+                    completed = suspensionPoint.completeExceptionally(failure);
+                }
+                if (!completed) {
+                    if (failure == null) {
+                        LOG.warn("Dropped completion");
+                    } else {
+                        LOG.warn("Dropped completion with exception", failure);
+                    }
+                }
+                return;
+            }
+            InternalFilter filter = filters.get(index);
+            if (filter instanceof InternalFilter.After<?> after) {
+                if (!invokeAfter(after)) {
+                    // suspend
+                    return;
+                }
+            }
+        }
+    }
+
+    private <T> boolean invokeBefore(InternalFilter.Before<T> before) {
         // todo: better param binding, handle return value, errors
-        before.method().invoke(before.bean(), request);
-        return ExecutionFlow.just(request);
+        try {
+            before.method().invoke(before.bean(), request);
```

Comment:
It would be interesting to see how is it going to look with return values.

Reply:
ive added support